src/hashmap/concurrent_unordered_map.cuh

```diff
@@ -0,0 +1,691 @@
+                {
+                    atomicAdd( &m_collisions, 1 );
+                }
+            }
+#else
+            
+            #pragma omp critical
+            {
+                if ( m_equal( unused_key, tmp_it->first ) ) {
+                    hashtbl_values[hash_tbl_idx] = thrust::make_pair( x.first, x.second );
+                    it = tmp_it;
+                }
+            }
+#endif
+            hash_tbl_idx = (hash_tbl_idx+1)%hashtbl_size;
+        }
+        
+        return iterator( m_hashtbl_values,m_hashtbl_values+hashtbl_size,it);
+    }
+    */
+    
+    __forceinline__
+    __host__ __device__ const_iterator find(const key_type& k ) const
+    {
+        size_type key_hash = m_hf( k );
+        size_type hash_tbl_idx = key_hash%m_hashtbl_size;
+        
+        value_type* begin_ptr = 0;
+        
+        size_type counter = 0;
+        while ( 0 == begin_ptr ) {
+            value_type* tmp_ptr = m_hashtbl_values + hash_tbl_idx;
+            const key_type tmp_val = tmp_ptr->first;
+            if ( m_equal( k, tmp_val ) ) {
+                begin_ptr = tmp_ptr;
+                break;
+            }
+            if ( m_equal( unused_key , tmp_val ) || counter > m_hashtbl_size ) {
+                begin_ptr = m_hashtbl_values + m_hashtbl_size;
+                break;
+            }
+            hash_tbl_idx = (hash_tbl_idx+1)%m_hashtbl_size;
+            ++counter;
+        }
+        
+        return const_iterator( m_hashtbl_values,m_hashtbl_values+m_hashtbl_size,begin_ptr);
+    }
+    
+    void assign_async( const concurrent_unordered_map& other, cudaStream_t stream = 0 )
+    {
+        m_collisions = other.m_collisions;
+        if ( other.m_hashtbl_size <= m_hashtbl_capacity ) {
+            m_hashtbl_size = other.m_hashtbl_size;
+        } else {
+            m_allocator.deallocate( m_hashtbl_values, m_hashtbl_capacity );
+            m_hashtbl_capacity = other.m_hashtbl_size;
+            m_hashtbl_size = other.m_hashtbl_size;
+            
+            m_hashtbl_values = m_allocator.allocate( m_hashtbl_capacity );
+        }
+        CUDA_RT_CALL( cudaMemcpyAsync( m_hashtbl_values, other.m_hashtbl_values, m_hashtbl_size*sizeof(value_type), cudaMemcpyDefault, stream ) );
+    }
+    
+    void clear_async( cudaStream_t stream = 0 ) 
+    {
+        constexpr int block_size = 128;
+        init_hashtbl<<<((m_hashtbl_size-1)/block_size)+1,block_size,0,stream>>>( m_hashtbl_values, m_hashtbl_size, unused_key, m_unused_element );
+        if ( count_collisions )
+            m_collisions = 0;
+    }
+    
+    unsigned long long get_num_collisions() const
+    {
+        return m_collisions;
+    }
+    
+    void print()
+    {
+        for (size_type i = 0; i < m_hashtbl_size; ++i) 
+        {
+            std::cout<<i<<": "<<m_hashtbl_values[i].first<<","<<m_hashtbl_values[i].second<<std::endl;
+        }
+    }
+    
+    void prefetch( const int dev_id, cudaStream_t stream = 0 )
```

Comment:
Is this prefetch only necessary because we used uvm before?

Reply:
This prefetch function is provided because the map class itself is allocated with managed memory and one may wish to prefetch the class. Furthermore, the hash table array (m_hashtbl_values) is optionally allocated with managed memory, in which case this function will prefetch that as well.